### PR TITLE
otpilot: Add flash support.

### PR DESCRIPTION
**Pull request by osenft**
This adds the userspace driver for H1's flash capsule.

### CODE REVIEW

File: userspace/otpilot/src/flash.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+use libtock::syscalls::raw::yieldk;
+
+pub const MAX_BUFFER_LENGTH: usize = 128;
+
+pub trait Flash {
+    // Read from flash.
+    // offset: Location relative to flash start to read from. Must be word (4 bytes) aligned.
+    // buffer: Buffer for read data.
+    //         It must be: buffer.len() <= len <= MAX_BUFFER_LENGTH
+    // read_length: Number of bytes to read. Must be > 0 and a multiple of 4.
```
Comment (dann21) [inline userspace/otpilot/src/flash.rs]: Nit: comment (`read_length`) is not consistent with parameter name (`len`).

File: userspace/otpilot/src/flash.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+use libtock::syscalls;
+use libtock::syscalls::raw::yieldk;
+
+pub const MAX_BUFFER_LENGTH: usize = 128;
+
+pub trait Flash {
+    // Read from flash.
+    // offset: Location relative to flash start to read from. Must be word (4 bytes) aligned.
+    // buffer: Buffer for read data.
+    //         It must be: buffer.len() <= len <= MAX_BUFFER_LENGTH
```
Comment (dann21) [inline userspace/otpilot/src/flash.rs]: Shouldn't `len` be `<= buffer.len()`?

File: userspace/otpilot/src/flash.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    // read_length: Number of bytes to read. Must be > 0 and a multiple of 4.
+    fn read(&self, offset: usize, buffer: &mut[u8], len: usize) -> TockResult<()>;
+
+    // Write to flash.
+    // offset: Location relative to flash start to write to. Must be word (4 bytes) aligned.
+    // buffer: Buffer with data to write.
+    //         It must be: buffer.len() <= len <= MAX_BUFFER_LENGTH
+    // read_length: Number of bytes to write. Must be > 0 and a multiple of 4.
```
Comment (dann21) [inline userspace/otpilot/src/flash.rs]: This should be `write_length`, not `read_length`.  And, again, the parameter name (`len`) is inconsistent with this comment.

File: userspace/otpilot/src/flash.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    fn clear_operation(&self);
+}
+
+// Get the static Flash object.
+pub fn get() -> &'static dyn Flash {
+    get_impl()
+}
+
+const DRIVER_NUMBER: usize = 0x40040;
```
Comment (dann21) [inline userspace/otpilot/src/flash.rs:61]: Why does this magic number have this value?  Should there be a reference (comment) here?

**Review by dann21 — APPROVED** [review]
LGTM, as far as I understand it.

File: userspace/otpilot/src/flash.rs (hunk 4 above)
Comment (osenft) [reply userspace/otpilot/src/flash.rs:61]: I think this is just a bit of "tribal" Tock knowledge on where to look for this number. One would usually look for the board-specific implementation of Tock's `kernel::platform::Platform`, which for us is in `tock-on-titan/kernel/papa/src/main.rs`. This lists all the drivers supported for the particular board. From there you can find the magic numbers.

File: userspace/otpilot/src/flash.rs (hunk 1 above)
Comment (osenft) [reply userspace/otpilot/src/flash.rs]: Oh, good catch, thank you! Fixed.

File: userspace/otpilot/src/flash.rs (hunk 2 above)
Comment (osenft) [reply userspace/otpilot/src/flash.rs]: Oops ... Yeah, of course. Thanks! Fixed.

File: userspace/otpilot/src/flash.rs (hunk 3 above)
Comment (osenft) [reply userspace/otpilot/src/flash.rs]: Fixed.

**Pull request merged**